File: storage-aws/src/test/scala/net/kemitix/thorp/storage/aws/ListerTest.scala
package net.kemitix.thorp.storage.aws

import org.scalatest.FreeSpec

class ListerTest extends FreeSpec {

//  "list" - {
//    val bucket = Bucket.named("aBucket")
//    val prefix = RemoteKey.create("aRemoteKey")
//    "when no errors" - {
//      "when single fetch required" in {
//        val nowDate         = new Date
//        val key             = "key"
//        val etag            = "etag"
//        val expectedHashMap = Map(MD5Hash.create(etag) -> RemoteKey.create(key))
//        val expectedKeyMap  = Map(RemoteKey.create(key) -> MD5Hash.create(etag))
//        new AmazonS3ClientTestFixture {
//          (() => fixture.amazonS3Client.listObjectsV2)
//            .when()
//            .returns(_ => {
//              UIO.succeed(objectResults(nowDate, key, etag, truncated = false))
//            })
//          private val result = invoke(fixture.amazonS3Client)(bucket, prefix)
//          private val hashMap =
//            result.map(_.byHash).map(m => Map.from(m.asMap.asScala))
//          private val keyMap =
//            result.map(_.byKey).map(m => Map.from(m.asMap.asScala))
//          hashMap should be(Right(expectedHashMap))
//          keyMap should be(Right(expectedKeyMap))
//        }
//      }
//
//      "when second fetch required" in {
//        val nowDate = new Date
//        val key1    = "key1"
//        val etag1   = "etag1"
//        val key2    = "key2"
//        val etag2   = "etag2"
//        val expectedHashMap = Map(
//          MD5Hash.create(etag1) -> RemoteKey.create(key1),
//          MD5Hash.create(etag2) -> RemoteKey.create(key2)
//        )
//        val expectedKeyMap = Map(
//          RemoteKey.create(key1) -> MD5Hash.create(etag1),
//          RemoteKey.create(key2) -> MD5Hash.create(etag2)
//        )
//        new AmazonS3ClientTestFixture {
//
//          (() => fixture.amazonS3Client.listObjectsV2)
//            .when()
//            .returns(_ =>
//              UIO(objectResults(nowDate, key1, etag1, truncated = true)))
//            .noMoreThanOnce()
//
//          (() => fixture.amazonS3Client.listObjectsV2)
//            .when()
//            .returns(_ =>
//              UIO(objectResults(nowDate, key2, etag2, truncated = false)))
//          private val result = invoke(fixture.amazonS3Client)(bucket, prefix)
//          private val hashMap =
//            result.map(_.byHash).map(m => Map.from(m.asMap.asScala))
//          private val keyMap =
//            result.map(_.byKey).map(m => Map.from(m.asMap.asScala))
//          hashMap should be(Right(expectedHashMap))
//          keyMap should be(Right(expectedKeyMap))
//        }
//      }
//
//      def objectSummary(key: String, etag: String, lastModified: Date) = {
//        val objectSummary = new S3ObjectSummary
//        objectSummary.setKey(key)
//        objectSummary.setETag(etag)
//        objectSummary.setLastModified(lastModified)
//        objectSummary
//      }
//
//      def objectResults(nowDate: Date,
//                        key: String,
//                        etag: String,
//                        truncated: Boolean) = {
//        val result = new ListObjectsV2Result
//        result.getObjectSummaries.add(objectSummary(key, etag, nowDate))
//        result.setTruncated(truncated)
//        result
//      }
//
//    }
//    "when Amazon Service Exception" in {
//      val exception = new AmazonS3Exception("message")
//      new AmazonS3ClientTestFixture {
//        (() => fixture.amazonS3Client.listObjectsV2)
//          .when()
//          .returns(_ => Task.fail(exception))
//        private val result = invoke(fixture.amazonS3Client)(bucket, prefix)
//        assert(result.isLeft)
//      }
//    }
//    "when Amazon SDK Client Exception" in {
//      val exception = new SdkClientException("message")
//      new AmazonS3ClientTestFixture {
//        (() => fixture.amazonS3Client.listObjectsV2)
//          .when()
//          .returns(_ => Task.fail(exception))
//        private val result = invoke(fixture.amazonS3Client)(bucket, prefix)
//        assert(result.isLeft)
//      }
//    }
//    def invoke(amazonS3Client: AmazonS3Client.Client)(bucket: Bucket,
//                                                      prefix: RemoteKey) = {
//      object TestEnv extends Storage.Test with Console.Test
//      val program: RIO[Storage with Console, RemoteObjects] = Lister
//        .listObjects(amazonS3Client)(bucket, prefix)
//      val runtime = new DefaultRuntime {}
//      runtime.unsafeRunSync(program.provide(TestEnv)).toEither
//    }
//
//  }

}

File: storage-aws/src/test/java/net/kemitix/thorp/storage/aws/ListerTest.java
package net.kemitix.thorp.storage.aws;

import com.amazonaws.services.s3.model.ListObjectsV2Result;
import com.amazonaws.services.s3.model.S3ObjectSummary;
import net.kemitix.thorp.domain.*;
import org.assertj.core.api.WithAssertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.Map;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;

@ExtendWith(MockitoExtension.class)
public class ListerTest
        implements WithAssertions {

    private final Bucket bucket = Bucket.named("aBucket");
    private final RemoteKey prefix = RemoteKey.create("aRemoteKey");
    private final AmazonS3Client amazonS3Client;

    public ListerTest(@Mock AmazonS3Client amazonS3Client) {
        this.amazonS3Client = amazonS3Client;
    }

    @Test
    @DisplayName("single fetch")
    public void singleFetch() {
        //given
        Date nowDate = new Date();
        String key = "key";
        String etag = "etag";
        Map<MD5Hash, RemoteKey> expectedHashMap =
                Collections.singletonMap(
                        MD5Hash.create(etag),
                        RemoteKey.create(key)
                );
        Map<RemoteKey, MD5Hash> expectedKeyMap =
                Collections.singletonMap(
                        RemoteKey.create(key),
                        MD5Hash.create(etag)
                );
        given(amazonS3Client.listObjects(any()))
                .willReturn(objectResults(nowDate, key, etag, false));
        //when
        RemoteObjects result = invoke(amazonS3Client, bucket, prefix);
        //then
        assertThat(result.byHash.asMap()).isEqualTo(expectedHashMap);
        assertThat(result.byKey.asMap()).isEqualTo(expectedKeyMap);
    }

    @Test
    @DisplayName("two fetches")
    public void twoFetches() {
        //given
        Date nowDate = new Date();
        String key1 = "key1";
        String etag1 = "etag1";
        String key2 = "key2";
        String etag2 = "etag2";
        Map<MD5Hash, RemoteKey> expectedHashMap = new HashMap<>();
        expectedHashMap.put(
                MD5Hash.create(etag1),
                RemoteKey.create(key1));
        expectedHashMap.put(
                MD5Hash.create(etag2),
                RemoteKey.create(key2));
        Map<RemoteKey, MD5Hash> expectedKeyMap = new HashMap<>();
        expectedKeyMap.put(
                RemoteKey.create(key1),
                MD5Hash.create(etag1));
        expectedKeyMap.put(
                RemoteKey.create(key2),
                MD5Hash.create(etag2));
        given(amazonS3Client.listObjects(any()))
                .willReturn(objectResults(nowDate, key1, etag1, true))
                .willReturn(objectResults(nowDate, key2, etag2, false));
        //when
        RemoteObjects result = invoke(amazonS3Client, bucket, prefix);
        //then
        assertThat(result.byHash.asMap()).isEqualTo(expectedHashMap);
        assertThat(result.byKey.asMap()).isEqualTo(expectedKeyMap);
    }

    private ListObjectsV2Result objectResults(
            Date nowDate,
            String key,
            String etag,
            boolean truncated
    ) {
        ListObjectsV2Result result = new ListObjectsV2Result();
        result.getObjectSummaries().add(objectSummary(key, etag, nowDate));
        result.setNextContinuationToken("next token");
        result.setTruncated(truncated);
        return result;
    }

    private S3ObjectSummary objectSummary(
            String key,
            String etag,
            Date lastModified
    ) {
        S3ObjectSummary summary = new S3ObjectSummary();
        summary.setKey(key);
        summary.setETag(etag);
        summary.setLastModified(lastModified);
        return summary;
    }

    private RemoteObjects invoke(
            AmazonS3Client amazonS3Client,
            Bucket bucket,
            RemoteKey prefix
    ) {
        return S3Lister.lister(amazonS3Client)
                .apply(S3Lister.request(bucket, prefix));
    }
}
